File: Neetcode-150-js/Arrays and Hashing/longest-consecutive-sequence.js

// https://leetcode.com/problems/longest-consecutive-sequence/

var longestConsecutive = function(nums) {
    const len = nums.length;
    const visited = {};
    const indexes = {};

    for(let i = 0; i < nums.length; i++){
        visited[nums[i]] = false;
        indexes[nums[i]] = i;
    }
    
    let maxLength = 0;
    for(let i = 0; i < len; i++){
      let currentLength = 0;
      let temp = nums[i];
      let tempNum = temp;
      
      // forward propogation
      while(!visited[temp] && indexes[temp] !== undefined){
        currentLength++;
        visited[temp] = true;
        temp++;
      }
      
      // backward propogation
      temp = tempNum-1;
      while(!visited[temp] && indexes[temp] !== undefined){
        currentLength++;
        visited[temp] = true;
        temp--;
      }
      
      maxLength = Math.max(maxLength, currentLength);
    }
    
    return maxLength;
};

File: Striver Important Interview Questions/Day 4 - Arrays Part 4/longest-consecutive-sequence.cpp

// https://leetcode.com/problems/longest-consecutive-sequence/

class Solution {
public:
    int longestConsecutive(vector<int>& nums) {
        int maxLength = 0;
        int n = nums.size();
        if(n == 0) return maxLength;

        unordered_map<int, int> index;
        unordered_map<int, bool> visited;
        for(int i = 0; i < n; i++){
            index[nums[i]] = i;
            visited[nums[i]] = false;
        }

        for(int i = 0; i < n; i++){
            int tempNum = nums[i];
            int tempNumCopy = tempNum;
            int currLength = 0;
            // forward
            while(index.count(tempNum) > 0 && !visited[tempNum]){
                visited[tempNum] = true;
                currLength++;
                tempNum++;
            }
            // backward
            tempNum = tempNumCopy-1;
            while(index.count(tempNum) > 0 && !visited[tempNum]){
                visited[tempNum] = true;
                currLength++;
                tempNum--;
            }
            maxLength = max(maxLength, currLength);
        }

        return maxLength;
    }
};